spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,240 @@
+   * It checks if the catalog type matches either `SparkSessionCatalog` or `SparkCatalog` and
+   * collects their names.
+   *
+   * @param sparkConf The SparkConf object containing Spark configurations.
+   * @return A Set of Strings representing the names of Iceberg catalogs.
+   */
+  private Set<String> extractIcebergCatalogNames(SparkConf sparkConf) {
+    return Arrays.stream(sparkConf.getAllWithPrefix(SPARK_SQL_CATALOG_PREFIX))
+        .filter(
+            entry -> {
+              String val = entry._2();
+              return val.contains(SparkSessionCatalog.class.getName())
+                  || val.contains(SparkCatalog.class.getName());
+            })
+        .map(Tuple2::_1)
+        .collect(Collectors.toSet());
+  }
+
+  @Override
+  public boolean delegationTokensRequired(SparkConf sparkConf, Configuration hadoopConf) {
+    return !getRequireTokenCatalogs(sparkConf).isEmpty();
+  }
+
+  private Set<String> getRequireTokenCatalogs(SparkConf sparkConf) {
+    return extractIcebergCatalogNames(sparkConf).stream()
+        .filter(catalog -> checkDelegationTokensRequired(sparkConf, catalog))
+        .collect(Collectors.toSet());
+  }
+
+  private boolean checkDelegationTokensRequired(SparkConf sparkConf, String catalogName) {
+    String metastoreUri = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + URI_KEY, "");
+    String principal = sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + PRINCIPAL_KEY, "");
+    boolean isHiveType =
+        CATALOG_TYPE.equalsIgnoreCase(
+            sparkConf.get(SPARK_SQL_CATALOG_PREFIX + catalogName + TYPE_KEY, ""));
+    if (metastoreUri.isEmpty()
+        || principal.isEmpty()
+        || !isHiveType
+        || !UserGroupInformation.isSecurityEnabled()) {
+      return false;
+    }
+
+    try {
+      Credentials credentials = UserGroupInformation.getCurrentUser().getCredentials();
+      Token<?> currentToken = credentials.getToken(new Text(metastoreUri));
+      return currentToken == null;
+    } catch (IOException ex) {
+      LOG.error(
+          "Failed to get current user credentials for catalog {}: {}",
+          catalogName,
+          ex.getMessage(),
+          ex);
+      throw new RuntimeException(ex);
+    }
+  }
+
+  IMetaStoreClient createHmsClient(HiveConf conf) throws Exception {
+    return new HiveMetaStoreClient(conf, null, false);
+  }
+
+  @Override
+  public Option<Object> obtainDelegationTokens(
+      Configuration hadoopConf, SparkConf sparkConf, Credentials creds) {
+    Map<String, IMetaStoreClient> hmsClientCache = Maps.newHashMap();
+    Set<String> requireTokenCatalogs = getRequireTokenCatalogs(sparkConf);
+    LOG.debug("Require token Hive catalogs: {}", requireTokenCatalogs);
+    try {
+      for (String catalogName : requireTokenCatalogs) {
+        Optional<HiveConf> hiveConfOpt = buildHiveConf(sparkConf, hadoopConf, catalogName);
+        if (!hiveConfOpt.isPresent()) {
+          continue;
```

Comment:
When no valid hive conf but catalog is listed then it's an error which should blow up, right?